### PR TITLE
Support other storage providers starting with AWS S3.

### PR DESCRIPTION
**Pull request by jthomas**
WIP PR for #1

### CODE REVIEW

**Review by rabbah — COMMENTED** [review]
lgtm!

**Comment by jthomas** [comment]
This is now ready for review.

File: nimbella/storage/plugin_manager.py
Comment (joshuaauerbachwatson) [inline nimbella/storage/plugin_manager.py:18]: Note that we have a legacy ... customers whose storage creds do not include a provider field.   This should default to  GCS.   Not sure whether the handling for that case belongs here, just noting it.

File: nimbella/nimbella.py
Comment (joshuaauerbachwatson) [inline nimbella/nimbella.py:55]: The `creds.get` call can return `None` because customers who got their storage credentials before we had storage abstraction don't have a provider field.   This needs to default to the GCS provider.   Also noted elsewhere.

File: nimbella/nimbella.py
Comment (joshuaauerbachwatson) [inline nimbella/nimbella.py]: I'm not too worried about this assuming the actually occurring case of provider=None is handled.  It would mean that the field is present but corrupted.   But, perhaps there is _something_ we could do to indicate the nature of the problem for serviceability?

**Review by joshuaauerbachwatson — COMMENTED** [review]
You should not rely heavily on my review in terms of code correctness because I don't know Python.  If @rabbah doesn't have time to do a review, get another reviewer who knows Python, perhaps Michele or Nikhil to eyeball the code.

I did make comments re the need to deal with provider=None.   That must be addressed before merge.

Apparent coverage of the abstraction: excellent
Alignment with the node SDK: excellent
Choice of Python constructs: looks great to me, but what do I know
Test coverage: very good and much better than I did in the node SDK.
Overall, looks very good.

File: nimbella/nimbella.py
Comment (jthomas) [reply nimbella/nimbella.py:55]: The second argument to the `creds.get` method (`@nimbella/storage-gcs`) is a default value to be returned in case the Python dict does not have a property. This should handle the default case unless I've missed something?
https://stackoverflow.com/questions/9358983/dictionaries-and-default-values

Comment (joshuaauerbachwatson) [reply nimbella/nimbella.py:55]: Ooops.   You are right.   Should have noticed that.   Sorry.

File: nimbella/nimbella.py
Comment (jthomas) [reply nimbella/nimbella.py]: Thinking about this.... we should definitely log an error message to indicate the problem. Will add something.

**Review by rabbah — APPROVED** [review]
LGTM. 

I believe that a developer can use `.client` as an escape hatch to the underlying boto or gcp client. Is that right?

File: .github/workflows/publish-to-pypi.yml
Comment (rabbah) [inline .github/workflows/publish-to-pypi.yml:22]: I didn't see the GH run on this PR (there are no checks). Are they supposed to?
Also if there's an existing test coverage threshold, can we add it as the low bar and enforce it going forward/

**Comment by rabbah** [comment]
Last question: any changes required to the readme?

File: nimbella/nimbella.py
Comment (jthomas) [reply nimbella/nimbella.py]: I've added an exception to be thrown in this (very unlikely) case.

**Comment by jthomas** [comment]
> LGTM.
> 
> I believe that a developer can use `.client` as an escape hatch to the underlying boto or gcp client. Is that right?

Correct!

File: .github/workflows/publish-to-pypi.yml
Comment (jthomas) [reply .github/workflows/publish-to-pypi.yml:22]: Good point - I've now added the option to run the CI/CD pipeline on PRs (as well as commits). It will only run the final stage in the pipeline (publish to PyPi) if commit is for a release tag.

**Comment by jthomas** [comment]
> Last question: any changes required to the readme?

Yes - I'll update the overview of the API. I have got an issue open (#3) to add more comprehensive docs in future.

**Comment by rabbah** [comment]
@jthomas thanks for the updates and answering the questions. LGTM and I see the GHA ran successfully.
I re-approved the PR. Feel free to merge and fix up the commit message as you see fit.

**Review by rabbah — APPROVED** [review]

**Comment by jthomas** [comment]
I've added an overview of the new API to the README. This is now ready for merging.

**Pull request merged**
